### ISSUE TITLE
 JtR on Windows: OpenCL DLL usage

### ISSUE THREAD

**Issue by dwendt** [openwall/john#3191]
This bug initially started with `clGetPlatformIDs` returning `-1001` which is  `CL_PLATFORM_NOT_FOUND_KHR` and john dies which is wrong/bad. 

hashcat on the same system detects the platform fine using literally the same function. The only conclusion I can come to is that hashcat is doing something correctly in their build, and john is not. 

### Steps to reproduce
1. grab the first build of bleeding-jumbo you can find on google http://openwall.info/wiki/_media/john/johntheripper-v1.8.0.12-jumbo-1-bleeding-e6214ceab-2018-02-07-win-x64.7z
2. install a graphics card whose platform is opencl 1.2 
3. run `john.exe --list=opencl-devices` with a card that's restricted to OCL 1.2 platform (this should work fine with 2.1/2.2 headers...)

### System configuration
windows enterprise 10.0.16299.19, 64-bit
```
H:\john180j1w\bleeding>.\run\john.exe --verbosity=5 --list=opencl-devices
Warning: '/dev/shm' does not exists or is not a directory.

POSIX shared memory objects require the existance of this directory.
Create the directory '/dev/shm' and set the permissions to 01777.
For instance on the command line: mkdir -m 01777 /dev/shm
initUnicode(UNICODE, UTF-8/ISO-8859-1)
UTF-8 -> UTF-8 -> UTF-8
Error: No OpenCL-capable platforms were detected by the installed OpenCL driver.
Throw clError: clGetPlatformIDs() = -1001
Error: No OpenCL-capable devices were detected by the installed OpenCL driver.


H:\john180j1w\bleeding>.\run\john.exe --list=build-info
Warning: '/dev/shm' does not exists or is not a directory.

POSIX shared memory objects require the existance of this directory.
Create the directory '/dev/shm' and set the permissions to 01777.
For instance on the command line: mkdir -m 01777 /dev/shm
Version: 1.8.0.12-jumbo-1-bleeding-e6214ceab 2018-02-07 17:56:37 +0100
Build: cygwin 64-bit x86_64 SSE4.2 AC MPI + OMP
SIMD: SSE4.1, interleaving: MD4:3 MD5:3 SHA1:1 SHA256:1 SHA512:1
CPU tests: SSE4.2
$JOHN is ./run/
Format interface version: 14
Max. number of reported tunable costs: 4
Rec file version: REC4
Charset file version: CHR3
CHARSET_MIN: 1 (0x01)
CHARSET_MAX: 255 (0xff)
CHARSET_LENGTH: 24
SALT_HASH_SIZE: 1048576
Max. Markov mode level: 400
Max. Markov mode password length: 30
gcc version: 6.4.0
OpenCL headers version: 2.2
Crypto library: OpenSSL
OpenSSL library version: 0100020ef
OpenSSL 1.0.2n  7 Dec 2017
GMP library version: 6.1.2
File locking: fcntl()
fseek(): fseek
ftell(): ftell
fopen(): fopen
memmem(): System's
```
```
H:\hashcat-4.0.1>hashcat64.exe -I
hashcat (v4.0.1) starting...

OpenCL Info:

Platform ID #1
  Vendor  : NVIDIA Corporation
  Name    : NVIDIA CUDA
  Version : OpenCL 1.2 CUDA 9.1.84

  Device ID #1
    Type           : GPU
    Vendor ID      : 32
    Vendor         : NVIDIA Corporation
    Name           : GeForce GTX 970
    Version        : OpenCL 1.2 CUDA
    Processor(s)   : 13
    Clock          : 1253
    Memory         : 1024/4096 MB allocatable
    OpenCL Version : OpenCL C 1.2
    Driver Version : 391.01
```

## root cause
* `cygOpenCL-1.dll` is really bad. JTR should never use that, and instead prefer OpenCL.dll in the loadlibrary path.
* replacing it with OpenCL.dll installed to system32 by the gfx card driver makes everything *almost* work.
* somehow, this hashcat release made hashcat prefer the library load path DLL over `cygOpenCL-1.dll`: https://hashcat.net/forum/thread-6187-post-32962.html
* if JtR implements the same change, it should prevent this from happening.

## almost working after replacing the junk cygopencl DLL
* so it sort of works. the device is listed successfully. however, there's another problem from that point on. for some reason the new DLL can't find the headers in the kernel folder even though they are (my guess is cygwin vs windows path formats causing problems?)
```
PS H:\john180j1w\bleeding\run> ./john.exe --verbosity=5 --format=dmg-opencl --wordlist=H:\rockyou.txt example.txt
Warning: '/dev/shm' does not exists or is not a directory.

POSIX shared memory objects require the existance of this directory.
Create the directory '/dev/shm' and set the permissions to 01777.
For instance on the command line: mkdir -m 01777 /dev/shm
initUnicode(UNICODE, UTF-8/ISO-8859-1)
UTF-8 -> UTF-8 -> UTF-8
Device 0@konata: GeForce GTX 970
Using default input encoding: UTF-8
Loaded 1 password hash (dmg-opencl, Apple DMG [PBKDF2-SHA1 OpenCL 3DES/AES])
Cost 1 (iteration count) is 270270 for all loaded hashes
Will run 8 OpenMP threads
Loaded 7 hashes with 7 different salts to test db from test vectors
Options used: -I /run/kernels -cl-mad-enable -DSM_MAJOR=5 -DSM_MINOR=2 -cl-nv-verbose -D__GPU__ -DDEVICE_INFO=262162 -DSIZEOF_SIZE_T=8 -DDEV_VER_MAJOR=391 -DDEV_VER_MINOR=1 -D_OPENCL_COMPILER -DHASH_LOOPS=753 -DOUTLEN=32 -DPLAINTEXT_LENGTH=64 -DV_WIDTH=1 $JOHN/kernels/pbkdf2_hmac_sha1_kernel.cl
Build log: <kernel>:46:10: fatal error: 'opencl_device_info.h' file not found
#include "opencl_device_info.h"
         ^

Error -11 building kernel $JOHN/kernels/pbkdf2_hmac_sha1_kernel.cl. DEVICE_INFO=262162
OpenCL CL_BUILD_PROGRAM_FAILURE (-11) error in common-opencl.c:1111 - clBuildProgram failed.
```

**Comment by dwendt** [comment]
so, there's a pretty funny workaround for the path thing that makes it work 100% correctly with the official OpenCL.dll supplied by driver manufacturers. 

1. `cd` into ./run/kernels
2. `../john.exe whatever-arguments-you-want-here`
3. everything should work fine!

**Comment by claudioandre** [comment]
If the `nvopencl.dll` exists at the specified location (and you are in the mood), could you try this build?
* https://ci.appveyor.com/project/claudioandre/johntheripper/build/job/9ilux27hrglbt31v/artifacts
* I expect it to work out of the box

Download and extract **win_x64.zip** 
* if you need it, you can find the SHA256 hash of the zip file inside the `Console` tab [1].

[1] 7FD1A4618FD3462A7FF3AA03A52F87FDA61FEB1F49BC4A77F1C434DAE7DFE64F

**********
In case you want to see what is that, you can read:
* https://github.com/claudioandre/packages/blob/master/john-the-ripper/readme.md#windows

**Comment by claudioandre** [comment]
> Waiting news from you. Still, thinking out loud

I do not have (or had) any NVIDIA or Intel hardware. So, I never tested it properly. Anyway, I did test the **win_x64.zip** package seen above using an AMD GPU on Win 64 bits and I noticed that:
*  `cygOpenCL-1.dll` is doing exactly what the authors said it should do;
* when one is running inside CygWin, some **Nix* configuration have to be done **by the packager**
* otherwise, it wont be able to find the installed platforms (odd environment, Linux on Windows)

**Comment by dwendt** [comment]
It does work, out of the box. Was it due to an older version of `cygOpenCL-1.dll`?

* `nvopencl.dll` does exist, and `C:\Windows\System32\OpenCL.dll` is what would load it when replacing cygOpenCL-1.dll with it.

```
PS H:\john\1.8J1-bc1bbc96f> .\run\john.exe --list=opencl-devices
Platform #0 name: NVIDIA CUDA, version: OpenCL 1.2 CUDA 9.1.84
    Device #0 (0) name:     GeForce GTX 970
    Device vendor:          NVIDIA Corporation
    Device type:            GPU (LE)
    Device version:         OpenCL 1.2 CUDA
    Driver version:         391.01 [recommended]
    Native vector widths:   char 1, short 1, int 1, long 1
    Preferred vector width: char 1, short 1, int 1, long 1
    Global Memory:          4 GB
    Global Memory Cache:    208 KB
    Local Memory:           48 KB (Local)
    Constant Buffer size:   64 KB
    Max memory alloc. size: 1024 MB
    Max clock (MHz):        1253
    Profiling timer res.:   1000 ns
    Max Work Group Size:    1024
    Parallel compute cores: 13
    CUDA cores:             1664  (13 x 128)
    Speed index:            2084992
    Warp size:              32
    Max. GPRs/work-group:   65536
    Compute capability:     5.2 (sm_52)
    Kernel exec. timeout:   yes
    PCI device topology:    01:00.0

PS H:\john\1.8J1-bc1bbc96f> .\run\john.exe --list=build-info
Version: 1.8.0.12-jumbo-1-bleeding-bc1bbc96f 2018-03-07 07:58:55 -0300
Build: cygwin 64-bit x86_64 AVX AC OMP
SIMD: AVX, interleaving: MD4:3 MD5:3 SHA1:1 SHA256:1 SHA512:1
CPU tests: AVX
CPU fallback binary: john-sse41
OMP fallback binary: john-avx-non-omp
$JOHN is /run/
Format interface version: 14
Max. number of reported tunable costs: 4
Rec file version: REC4
Charset file version: CHR3
CHARSET_MIN: 1 (0x01)
CHARSET_MAX: 255 (0xff)
CHARSET_LENGTH: 24
SALT_HASH_SIZE: 1048576
Max. Markov mode level: 400
Max. Markov mode password length: 30
gcc version: 6.4.0
OpenCL headers version: 2.2
Crypto library: OpenSSL
OpenSSL library version: 0100020ef      (loaded: 0100020df)
OpenSSL 1.0.2n  7 Dec 2017      (loaded: OpenSSL 1.0.2m  2 Nov 2017)
GMP library version: 6.1.2
File locking: fcntl()
fseek(): fseek
ftell(): ftell
fopen(): fopen
memmem(): System's
```

**Comment by dwendt** [comment]
Well, there's still strange path problems going on. The previous workaround for that still works, though.
```
PS H:\john\1.8J1-bc1bbc96f> ./run/john.exe --verbosity=5 --format=dmg-opencl --wordlist=H:\rockyou.txt H:\example.txt
initUnicode(UNICODE, UTF-8/ISO-8859-1)
UTF-8 -> UTF-8 -> UTF-8
Device 0: GeForce GTX 970
Using default input encoding: UTF-8
Loaded 1 password hash (dmg-opencl, Apple DMG [PBKDF2-SHA1 3DES/AES OpenCL])
Cost 1 (iteration count) is 270270 for all loaded hashes
Cost 2 (version) is 2 for all loaded hashes
Loaded 7 hashes with 7 different salts to test db from test vectors
Options used: -I /run/kernels -cl-mad-enable -DSM_MAJOR=5 -DSM_MINOR=2 -cl-nv-verbose -D__GPU__ -DDEVICE_INFO=262162 -DSIZEOF_SIZE_T=8 -DDEV_VER_MAJOR=391 -DDEV_VER_MINOR=1 -D_OPENCL_COMPILER -DHASH_LOOPS=500 -DOUTLEN=32 -DPLAINTEXT_LENGTH=64 -DV_WIDTH=1 $JOHN/kernels/dmg_kernel.cl
Build log: <kernel>:14:10: fatal error: 'pbkdf2_hmac_sha1_kernel.cl' file not found
#include "pbkdf2_hmac_sha1_kernel.cl"
         ^

Error -11 building kernel $JOHN/kernels/dmg_kernel.cl. DEVICE_INFO=262162
OpenCL CL_BUILD_PROGRAM_FAILURE (-11) error in opencl_common.c:1115 - clBuildProgram failed.
```

**Comment by claudioandre** [comment]
> It does work, out of the box. Was it due to an older version of cygOpenCL-1.dll?

No. It was caused because `cygOpenCL-1.dll` needs a config file [1] that points to `nvopencl.dll` [2].


> Well, there's still strange path problems going on. The previous workaround for that still works, though.

But, have you replaced `cygOpenCL-1.dll` with `OpenCL.dll`? 
* If you did that, the error is expected to happen. JtR is not ready to handle a real Windows OpenCL dll.

**********
[1] the ICD vendor file.
[2] well, since I don't have NVIDIA, I asked you to check if the file really exists, or if my (cof, Google) guess was wrong.

**Comment by dwendt** [comment]
Nope, the DLL isn't replaced. This is just regularly unzipping that build to a folder, and running that cmdline.

**Comment by claudioandre** [comment]
Ok. We need to think about how to fix it.

**Comment by claudioandre** [comment]
Could you please try https://ci.appveyor.com/project/claudioandre/johntheripper/build/1.8J1-688d3d200/job/frq42583ylqvuhwx/artifacts ?

Important:
* you have to delete all `*.bin`files that exist inside `run\kernels` (to force a rebuild).
* you will see some debug output (example below).

Just an example (run it the way you used to)
```
C:\bleeding>del \bleeding\run\kernels\*.bin && \bleeding\run\john --test=0 --format=sha512crypt-opencl
Device 0: Juniper [AMD Radeon HD 6700 Series]
Testing: sha512crypt-opencl, crypt(3) $6$ (rounds=5000) [SHA512 OpenCL]... ---
Testing: ./run/kernels
Final: ./run/kernels
Where: /
PASS
```

**Comment by claudioandre** [comment]
**Ops, the link I used was for the 32-bit binary. Excuse me.**

Fixed.

**Comment by arcfide** [comment]
I'm having trouble using OpenCL on Windows with stuff like this. I'm confused reading the thread above what the correct process is to getting this to work? I've read about ICD vendor files for Cygwin, and I think I understand what I need to put in the ICD file to get things to work, but I can't see where I would put the ICD vendor file. Where in the JTR unzipped program folder should I put an ICD vendor file?

**Comment by kholia** [comment]
> Where in the JtR unzipped program folder should I put an ICD vendor file?

This thread doesn't talk about ICD vendor file(s) at all, if I am reading things correctly.

I will be ("soon" -> no real eta) working on making sure that JtR's OpenCL support just runs on Windows without any hard hacks.

**Comment by kholia** [comment]
@claudioandre-br Hey, do you already have some AppVeyor configuration (appveyor.yml) to build JtR Jumbo using Cygwin on Windows? If yes, could you please open a PR to include this build configuration in our official repository?

This way, I can build on top of your work and don't have to start from scratch.

**Comment by arcfide** [comment]
Okay, I got this fixed up. If you see claudioandre-br's comment, that's where ICD Vendor files are mentioned. He also gives a working example of a build that seems to work. I've got this working now on the current build.

It doesn't require any hard hacks, but I did figure out that the OpenCL drivers with Cygwin don't work without an ICD Vendor file. That means that there has to be a location to find such files. That means that the OpenCL support works on Windows if you run JtR from inside of a Cygwin installation. 

To make this work for me on Windows, I installed Cygwin with OpenCL, and then created the /etc/OpenCL/vendors/nvidia.icd file that included the Cygwin path to nvopencl.dll mentioned above. After I did that, I ran JtR from inside of the Cygwin Terminal, which mounts and makes available the /etc/ directory. That has fixed things, and I can now see all of my devices and I can run JtR on the GPU with the appropriate speedups.

**Comment by claudioandre-br** [comment]
> OpenCL drivers with Cygwin don't work without an ICD Vendor file.

Indeed

> To make this work for me on Windows, I installed Cygwin with OpenCL [...] I ran JtR from inside of the Cygwin Terminal

It it possible to run it without Cygwin installed. Using this zip file (https://rebrand.ly/JtRWin64), you can run JtR on a regular Windows machine (GPU drivers required, or course). And it is 100% magnum ripper source code.

**Comment by kholia** [comment]
Thanks for these hints. 

I am looking forward to create redistributable JtR-Jumbo-OpenCL builds which just work without requiring installation of a Cygwin environment. Maybe it is not possible to do what I am planning to do? I don't know yet.

**Comment by kholia** [comment]
> @claudioandre-br Hey, do you already have some AppVeyor configuration (appveyor.yml) to build JtR Jumbo using Cygwin on Windows? If yes, could you please open a PR to include this build configuration in our official repository?

I see that you have https://github.com/claudioandre-br/packages/blob/master/john-the-ripper/tests/appveyor.yml.

Could it be made part of the official JtR Jumbo repository? Can we get the latest Windows + Cygwin builds to test using this method? If yes, this will greatly reduce a lot of friction for the end-users.

**Comment by claudioandre-br** [comment]
> do you already have some AppVeyor configuration (appveyor.yml) to build JtR 

Yes, but it was designed to test, not exactly to create a package.

> redistributable JtR-Jumbo-OpenCL builds which just work without requiring installation of a Cygwin environment. Maybe it is not possible to do what I am planning to do?

Yes, it is. It did this in https://rebrand.ly/JtRWin64
In my process, I'm building it as a special since my process was designed to test JtR:
* the JtR package is 100% magnum ripper code;
* the JtR package do not run all tests, the goal is to create the package.
* my regular build (not the build that creates JtR package) disable cursed formats and run all tests;

> Could it be made part of the official JtR Jumbo repository?

Yes

**Comment by claudioandre-br** [comment]
Just in case anyone arrives here by chance, an introductory text:
* https://github.com/claudioandre-br/packages/blob/master/john-the-ripper/readme.md#windows

I'm closing the bug. It is done.

**Comment by kholia** [comment]
>> Could it be made part of the official JtR Jumbo repository?

> Yes. But I created this file to test. It might not be suitable to be used as you intend.

I see. Nevertheless, it would be great to include this file in the official source tree. It currently does JtR Jumbo testing and with some future work, we can probably even generate ready-to-use builds using this AppVeyor configuration file.

> https://github.com/claudioandre-br/packages/blob/master/john-the-ripper/readme.md#windows

This file contains great information :+1: Some of the content here belongs in the official source tree.

**Comment by claudioandre-br** [comment]
Anything can be copied and used. Including instructions and the https://rebrand.ly/JtRWin64 link.

I'm not going to do this myself because a user can report as an issue:
```
Hey magnum

It installed JtR using the link https://rebrand.ly/JtRWin64. Two weeks later, I accessed 
www.OMGosh-photos.com and my computer was hacked. What should I do?
```

I bet magnum will say that packaging is something OpenWall should do.

**Comment by kholia** [comment]
Thanks. I can do a few PR(s) regarding this - hopefully not too far away in the future.

We already build, test, and provide Win32 MinGW builds using the official repository. These builds aren't perfect (one issue -> AVX2 missing on 64-bit builds due to GCC bug) but they are better than nothing.

With some more work (and more documentation), these MinGW builds can be made even more usable.

With your work on Cygwin builds, we will have even more usable and featureful builds in the near future :+1:

All this needs to be part of the official source tree, else it will bitrot sooner or later.

@solardiz @magnumripper I would like to hear your opinions on this topic of creating usable builds for end-users. Would doing so be a desirable thing?

I know that @magnumripper is concerned about us having to provide support for such builds but if we set the right expectations with our users (one idea -> GitHub issue template will ask users to post to john-users regarding support for any binary builds instead of opening a issue in the repository), this shouldn't be a big problem. It will be a win for us (as developers) as we would get to iron out a few bugs in our code base, I imagine.

**Comment by magnumripper** [comment]
> I bet magnum will say that packaging is something OpenWall should do.

Well, either that or no-one. I certainly wont.

**Comment by technowhiz** [comment]
Hi, I am a newbie to john the ripper. I am getting the opencl error which is specified here. could anyone explain me how to replace the cygOpencl-1.dll with openCL.dll in the john the ripper directory. I copied the OpenCL.dll from system32 folder to the 'run' directory inside the john the ripper directory but got the same opencl error. then i renamed the OpenCL.dll to cygOpenCL-1.dll. now i am getting cmd error "unable to start application 0xc000007b". kindly help me to solve this issue.
thanks in advance

Edit: dropped the extreme over-quoting.

**Comment by solardiz** [comment]
@sdk156 We use GitHub issues to track things for us to fix in JtR, not to provide user support. Your question would be appropriate for the john-users mailing list: https://www.openwall.com/lists/john-users/

Then, when you use the latest Windows build currently on JtR homepage (of version 1.9.0-jumbo-1), the problem shouldn't occur. You didn't mention what version you have the problem with (you should; and if it's not the latest, then we won't care).

Finally, please avoid over-quoting. Instead of the quote, you should have provided actual error output you're getting.

Summary: start by trying out the latest version/build off JtR homepage. If you still have issues with it, join the john-users mailing list and post in there (and mention what version/build of JtR you try to use). Thanks!

**Comment by solardiz** [comment]
@sdk156 You should also mention the type of your GPU, what GPU drivers you installed (if any), and Windows version and bitness.

Reviewing past comments on this issue, I see @arcfide mentioning that things work with Cygwin's DLL when ICD files are added. That's similar to what we did in our currently released builds, where we included ICD files with typical paths to NVIDIA and AMD DLLs. (But not to Intel initially - that's fixed in Claudio's later builds.)

**Comment by technowhiz** [comment]
@solardiz my aplologies for the incomplete error report. I have a NVIDIA Geforce MX150 GPU with latest NVIDIA drivers installed. And I was getting the same error as listed below.
Error: No OpenCL-capable platforms were detected by the installed OpenCL driver.
Error: No OpenCL-capable devices were detected by the installed OpenCL driver.

the releases i was trying as follows,
1. winX64_1_JtR.7z - latest dev version
2. john-1.9.0-jumbo-1-win64 - latest windows binary from openwall/john
3.johntheripper-v1.8.0.12-jumbo-1-bleeding-e6214ceab-2018-02-07-win-x64.7z - as suggested in this thread
4.x64_win.zip

and fortunately, i have resolved the issue by replacing the contents of nvidia.icd under 'etc/OpenCL/vendors'. The default contents of the file is 'c:\Windows\System32\nvopencl.dll'. I have replaced this with the actual path of nvopencl64.dll in another path under system32 folder. Now jon the ripper successfully listed my gpu when using the command 'john --list=opencl-devices'

**Comment by solardiz** [comment]
Thank you for this additional detail, @sdk156.

What was the actual path of nvopencl64.dll on your system? Do you know why it's different from the default?

I wonder if we can somehow list multiple/alternate paths, @claudioandre-br.

**Comment by technowhiz** [comment]
@solardiz , sorry . I am a noob to opencl environments. In this thread , I have read that nvopencl.dll might be found under system32 folder. so i just searched it inside system32 but it was not under system32 root but under a subdirectory as follows 'C:\Windows\System32\DriverStore\FileRepository\nvdmi.inf_amd64_b67df4d626c9e7b6'. so i replaced this path instead of the contents of nvidia.icd  in 'etc/OpenCL/vendors'. And it worked. Once again sorry, i dont know why i have a different path for nvopencl64.dll.
thank you.

**Comment by solardiz** [comment]
We appreciate the detailed info you provided, @sdk156. Thank you!

**Comment by technowhiz** [comment]
@solardiz welcome and thank you for your appreciation :) :) :)

**Comment by samyH** [comment]
Hi guys,

I'm having issues getting my GPU (Nvidia GTX 860M) to even output as a device when running `./john --list=opencl-devices`.

The driver I currently have installed (via Device Manager) is 23.21.13.8813 - I have tried rolling back to drivers from 2014 and also updating to 2019 driver with no luck.

I am using Cygwin64 with john-1.9.0-jumbo-1-win64 build from openwall website.

I also tried changing the path inside the nvidia.icd vendor file to the location of nvopencl64.dll as mentioned above, but no luck.

I know it's not the fastest GPU in the world, but hopefully, it'll help me get to my password a little quicker if I can get this working correctly.

Thanks in advance.

**Comment by solardiz** [comment]
Dear all, let me inform/remind you that our use of GitHub is to track issues for us to fix. It is not to provide user support. We have the john-users mailing list for that. So if you post something worded as a support request on our GitHub, you're doing it wrong. ;-)

> I also tried changing the path inside the nvidia.icd vendor file to the location of nvopencl64.dll as mentioned above, but no luck.

@samyH Did you find where nvopencl64.dll is located on _your_ system and try using that path? There's almost no chance @sdk156's cryptic path would also be right for your system.

**Comment by technowhiz** [comment]
@samyH ,  I think there is no need to use cygwin in windows to run JTR. And i read in one of the forums that, windows still does not support gpu access via its windows subsytem for linux (WSL). plz  follow  this [WSL gpu issue](https://github.com/Microsoft/WSL/issues/829). this issue is still open. Try replacing the path of nvopencl64.dll with your own and  try running JTR in command prompt.

**Comment by solardiz** [comment]
@claudioandre-br @alainesp Can we possibly extract the vendors' OpenCL backend DLL pathnames from somewhere in Windows and populate (e.g., from JtR startup code) the .icd file(s) for Cygwin's cygOpenCL-1.dll that our builds use? Where is this information normally stored on Windows - e.g., registry or does it have .icd file(s) somewhere?

**Comment by alainesp** [comment]
Hi Solar.

On Wed, Jul 3, 2019 at 7:46 AM Solar Designer <notifications@github.com>
wrote:

> @claudioandre-br <https://github.com/claudioandre-br> @alainesp
> <https://github.com/alainesp> Can we possibly extract the vendors' OpenCL
> backend DLL pathnames from somewhere in Windows and populate the .icd
> file(s) for Cygwin's cygOpenCL-1.dll that our builds use? Where is this
> information normally stored on Windows - e.g., registry or does it have
> .icd file(s) somewhere?
>
This project may help  https://github.com/KhronosGroup/OpenCL-ICD-Loader
In the Readme.md they write:

On Windows, add the "stub" ICD by adding a REG_DWORD value to the registry
keys:

// For 32-bit operating systems, or 64-bit tests on a 64-bit operating system:
HKEY_LOCAL_MACHINE\SOFTWARE\Khronos\OpenCL\Vendors

// For 32-bit tests on a 64-bit operating system:
HKEY_LOCAL_MACHINE\SOFTWARE\Wow6432Node\Khronos\OpenCL\Vendors

// The name of the REG_DWORD value should be the full path to the "stub" ICD
// OpenCLDriverStub.dll, and the data for this value should be 0.


Hope this helps,
Alain

**Comment by solardiz** [comment]
Thanks Alain. So what we may do is _one_ of:

1. Read those registry keys and populate an .icd file for use by cygOpenCL-1.dll.

2. Load Windows' native OpenCL.dll from our code.

3. Can we just link to Windows' native OpenCL.dll instead of to Cygwin's somehow, so that we don't have to complicate our code with manual loading of the library?

4. As a last resort, maybe we can include a script or whatever that would do what Claudio suggests here:

https://github.com/claudioandre-br/packages/blob/master/john-the-ripper/readme.md#windows

"the first advice to be given to anyone facing Windows problems would be: replacing cygwin's OpenCL library cygOpenCL-1.dll with OpenCL.dll installed in the System32 folder should make everything almost work."

If that's in fact our advice, then we should make it available to users and easy to follow, not have it "hidden" in Claudio's repo.

I'll reopen the issue now. Claudio closed it with "I'm closing the bug. It was not reporting a bug." Bug or not, this is an issue that many people keep running into, and we ought to do something about that.

**Comment by solardiz** [comment]
The original issue description here says:

> somehow, this hashcat release made hashcat prefer the library load path DLL over `cygOpenCL-1.dll`: https://hashcat.net/forum/thread-6187-post-32962.html

I just found the corresponding hashcat commit: https://github.com/hashcat/hashcat/commit/33aeae60908e9dcdc9165dff5897d90d661ca9f6

It does basically:

```diff
   #elif defined (__CYGWIN__)
-  ocl->lib = hc_dlopen ("cygOpenCL-1.dll", RTLD_NOW);
+  ocl->lib = hc_dlopen ("opencl.dll", RTLD_NOW);
+
+  if (ocl->lib == NULL) ocl->lib = hc_dlopen ("cygOpenCL-1.dll", RTLD_NOW);
```

So hashcat was already using `dlopen` before that point. We don't yet.

**Comment by solardiz** [comment]
@dwendt Maybe you'd want to look into this issue further and contribute a fix to our project in the form of a pull request? We'd appreciate that! Most of us are not on Windows, but it appears that you are, and it appears that you'd also be capable of coming up with a proper fix for all, not just for your computer.

**Comment by dwendt** [comment]
Hey @solardiz - apologies for opening this and then orphaning it. I was working through this for a side project, totally unrelated to my day job, and I see that other people have apparently had the same issue...

Thanks for finding that change - I'm going to get john building and experiment on a fresh Windows install on both an AMD & Nvidia system to see if doing something similar solves it. I'll PR some time this weekend if it does.

**Comment by StillMyGitHub** [comment]
I had the same issue with john-1.9.0-jumbo-1-win64 and Radeon RX 580.
replacing the dll and cd-ing to \kernels\ did not work for me at first try, the funny thing I noticed is, that this does only work with relative adressing of john.exe. If you use absolute adressing (C:\Users\%Username%\....\john.exe) john cannot find the kernel headers even if it is started from that path.

Furthermore: After a first successful run which builds the kernel (a .bin file in the same kernels folder) you can run john (for that hashtype) from any directory

I wonder whether...
a) it is possible to just precompile all kernels 
b) how customized/ portable the OpenCL kernels are
c) why it makes a difference wether you start john with absolute or relative path

**Comment by magnumripper** [comment]
a) Pre-compiling is not possible right now. We might want to have that, especially for DEScrypt per-salt kernels (4K of them...).

b) It depends. Some vendors' binary kernels are tightly coupled to runtime/driver version and hardware. I believe some others aren't, they're more like a compile-as-far-as-you-can-without-getting-specific. I'm sure @solardiz knows more about it.

c) That's a jolly good question. I'm not sure I even want to hear the answer 🤣

**Comment by ghost** [comment]
>> c) why it makes a difference wether you start john with absolute or relative path
>
> c) That's a jolly good question. I'm not sure I even want to hear the answer

I was aware of that. When you run JtR from outside the CygWin root folder, it fails to resolve its "home" properly.
- the CygWin function call that retrieves the path fails.

**Comment by magnumripper** [comment]
So what can we do about that? Isn't there a good system call to use on Windows crap?

**Comment by solardiz** [comment]
On Wed, Oct 09, 2019 at 04:36:59PM -0700, Claudio Andr?? wrote:
> When you run JtR from outside the CygWin root folder, it fails to resolve its "home" properly.
> - the CygWin function call that retrieves the path fails.

If so, perhaps this affects more than just OpenCL kernels - e.g., does it also affect JtR's ability to find its john.conf and the CPU fallback binaries?  If somehow this issue does not affect those things (why not?) then perhaps we can revise our OpenCL support code to also avoid the issue in the same way as those other places do?

**Comment by ghost** [comment]
I can't remember. I would say:
* OpenCL kernels compilation lib call requires a full path. 
* OR, `real_path()` is failing (shouldn't be used).

Some Windows tests have to be run.

**Comment by ghost** [comment]
For the record ([1]):

My expectation was that a proprietary GPU driver installation on Windows would add OpenCL "instructions" to the registry at _HKEY_LOCAL_MACHINE\SOFTWARE\Khronos\OpenCL\Vendors_. Unfortunately, this is **not** a rule [2].

On my machine I did not find any OpenCL significant information within the registry. To make OpenCL work I did:
* copy the file `OpenCLdll` installed in the `c:\Windows\System32` folder to JtR's `run` directory, and rename it to `cygOpenCL-1.dll`.
* after that, JtR was able to detect all OpenCL devices.

I do not believe that there is a universal (mine, for example) `Openll.dll` that works for everyone.

That said, #2133 is a good idea for 2.0 series.

----
[1] I just purchased a new Acer "Ryzen" notebook.
[2[ The AMD GPU/OpenCL came installed from the "factory" (OEM).

**Comment by magnumripper** [comment]
I recently (do not hold this against me, I will forever deny it) helped someone to install JtR on some Windoes gear. I tried using your  (@claudioandre-br) latest binary and did just that - copied a known-good nvidia `OpenCL.dll` right over `cygOpenCL-1.dll`. Worked like a champ (but not until that). Good lord knows I was relieved.

**Comment by solardiz** [comment]
Given all of the above and the recent reminder in #5262, I wonder if we can workaround the issue by somehow linking against `OpenCL.dll` instead of `cygOpenCL.dll` (or is it `cygOpenCL-1.dll`). As I understand. right now we link against `-lOpenCL`, so the usage of `cygOpenCL.dll` (or `cygOpenCL-1.dll`) is added implicitly by Cygwin. Can we bypass that by usage of different options to the linker? If not, or not easily, then maybe we should add a step where we use e.g. `sed -i` to patch the library name in the already-linked binary? Luckily, `cygOpenCL.dll` is longer than `OpenCL.dll`, so we should be able to replace one string by the other plus several NUL bytes. @claudioandre-br or anyone else who's on Windows (maybe @boulevard-h from #5262 as someone who ran into this issue most recently?), can you please give this a try? Thank you!

**Comment by solardiz** [comment]
This seems to do what's needed, but I did not actually test on Windows (just checked that the file is changed the way I want, on two Linux systems with different versions of `sed` and different locales):

```shell
LC_CTYPE=C sed -rbi 's/cygOpenCL-1\.dll(.)/OpenCL.dll\1\1\1\1\1\1/' john.exe
```

Please test and let us all know. If this works as desired, then let's get it into our build procedure for Cygwin (perhaps into the `configure`-generated `Makefile` and into `Makefile.legacy`).

**Comment by ghost** [comment]
I'm afraid this is not a good solution as we also need to package the cygWin library (otherwise error in non-OpenCL environments). 

The example below will show where the devil lives:

```
C:\Temp\debug> run\john --test --format=nt-opencl
Device 1: gfx902 [AMD Radeon(TM) Vega 8 Graphics]
Benchmarking: NT-opencl [MD4 OpenCL/mask accel]... LWS=64 GWS=2048 (32 blocks) x2470 DONE
Raw:    1170M c/s real, 6911M c/s virtual

C:\Temp\debug> del run\cygOpenCL-1.dll

C:\Temp\debug> dir run\cyg*.dll
    [...]  
   8 file(s)      7.180.198 bytes

# => john useless
C:\Temp\debug> run\john --test --format=nt
C:/Temp/debug/run/john-best.exe: error while loading shared libraries: cygOpenCL-1.dll: cannot open shared object file: No such file or directory

C:\Temp\debug> run\john --help
C:/Temp/debug/run/john-best.exe: error while loading shared libraries: cygOpenCL-1.dll: cannot open shared object file: No such file or directory

C:\Temp\debug> copy c:\Windows\System32\OpenCL.dll c:\Windows\System32\cygOpenCL-1.dll
# Admin is required.
# => Now we have a valid and functional OpenCL library, please don't worry about the name.

C:\Temp\debug> run\john --test --format=nt-opencl
Device 1: gfx902 [AMD Radeon(TM) Vega 8 Graphics]
Benchmarking: NT-opencl [MD4 OpenCL/mask accel]... LWS=64 GWS=16384 (256 blocks) x2470 DONE
Raw:    1489M c/s real, 29880M c/s virtual
```

What if we have some cygwin OpenCL DLL in run folder:
```
#=> Copy the CygWin original cygOpenCL-1.dll to the run folder.

C:\Temp\debug> dir run\cyg*.dll
[...]
01/01/2018  09:55           110.099 cygOpenCL-1.dll
[...]               9 files(s)      7.290.297 bytes

C:\Temp\debug> run\john --test --format=nt-opencl
No OpenCL devices found
```

**Comment by solardiz** [comment]
> I'm afraid this is not a good solution as we also need to package the cygWin library (otherwise error in non-OpenCL environments).

I think I get what you mean, even though your examples don't really show it - this would matter for Windows systems that don't have `c:\Windows\System32\OpenCL.dll`, and it would prevent usage of our CPU formats, which otherwise should work even on such systems. (That `nt-opencl` fail, like in your examples, is of less importance - on such systems, it'd have failed anyway, just differently. However, if you ran e.g. `run\john --test --format=nt`, it'd also bump into the `error while loading shared libraries`, which would actually make `john` useless.)

Is `c:\Windows\System32\OpenCL.dll` in fact optional on modern Windows, or do modern versions of Windows provide it even when there are no OpenCL device drivers installed? Would it be sane for us to assume that any 64-bit Windows has this DLL?

**Comment by ghost** [comment]
> e.g. `run\john --test --format=nt`, it'd also bump into the error while loading shared libraries, which would actually make john useless.)

100% useless, not even a `run\john --help`.

> do modern versions of Windows provide it even when there are no OpenCL device drivers installed?

Well, I don't know, but I can uninstall stuff, so I would say that's a dangerous assumption. But we should call people to comment on the matter.

**[edited]**

In fact, we can use AWS Amazon EC2 T2 (Free Tier) to verify this. I'll do that when I have my security keys in hand.

**Comment by solardiz** [comment]
Skimming https://learn.microsoft.com/en-us/windows/win32/dlls/dynamic-link-library-search-order I see that "Known DLLs" are of higher priority than those in "The folder from which the application loaded." So _if_ `OpenCL.dll` is listed in `HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Control\Session Manager\KnownDLLs`, then we can get around the issue just brought up by @claudioandre-br above by shipping a copy of Cygwin's `cygOpenCL-1.dll` renamed to `OpenCL.dll` in the current directory, for it to be used as fallback when the system does not have its own copy of `OpenCL.dll`. OTOH, I don't know if it's commonly or ever the case that `OpenCL.dll` is listed among known ones even when it's missing?

If `OpenCL.dll` isn't "known", then unfortunately our program directory's version of it will take precedence over the system's, which is not what we want here.

**Comment by ghost** [comment]
> `HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Control\Session Manager\KnownDLLs`

Starting with 'O' only 2 files: `ole*.dll`.
****

Just to document it here, another way to fix it:
- Locate the vendor's OpenCL dll file. In my case it is [1]:
  `C:\WINDOWS\System32\DriverStore\FileRepository\u0357645.inf_amd64_3c66802dbd4d45a0\B357435\amdocl64.dll`
- (if you are using https://github.com/openwall/john-packages) Inside the main john folder you will find a subdirectory called `etc`
```
C:\Temp\debug> dir etc\OpenCL\vendors\*
[...]

 Folder C:\Temp\debug\etc\OpenCL\vendors

02/04/2023  18:25    <DIR>          .
02/04/2023  18:25    <DIR>          ..
02/04/2023  18:30               104 amd.icd
31/03/2023  13:20                38 intel.icd
31/03/2023  13:20                33 nvidia.icd
               3 arquivo(s)            175 bytes
               2 pasta(s)   321.126.096.896 bytes disponíveis

```

- Edit the `.icd` file and point it to the correct vendor DLL file. Just put the DLL full name inside the file.
```
# This is not valid Windows syntax, In fact, I used Notepad
echo 'C:\WINDOWS\System32\DriverStore\FileRepository\u0357645.inf_amd64_3c66802dbd4d45a0\B357435\amdocl64.dll' > etc\OpenCL\vendors\amd.icd
```


[1] A very good and easy to remember place to save a library.
      On Linux, we use something awful like `/lib` and a symbolic link.

**Comment by solardiz** [comment]
>* Edit the `.icd` file and point it to the correct vendor DLL file. Just put the DLL full name inside the file.

Apparently, we could automate that by reading the registry and writing to Cygwin's `.icd` file (we could do this in `john` or in a separate program, and would need to inform the user of us making such changes outside of our program tree). See: https://github.com/openwall/john/issues/3191#issuecomment-508086364

**Comment by alainesp** [comment]
What is the problem with:

In the configure script. If we are under Cygwin and OpenCL is supported:
 - copy the file C:\Windows\System32\OpenCL.dll to JtR's run directory, renaming it to cygOpenCL-1.dll
 -  If no file exists we print a warning/error

**Comment by solardiz** [comment]
> In the configure script. If we are under Cygwin and OpenCL is supported:
> 
>     * copy the file C:\Windows\System32\OpenCL.dll to JtR's run directory, renaming it to cygOpenCL-1.dll
> 
>     * If no file exists we print a warning/error

This only works for those building from source on the very machine where they intend to run. Most Windows users will probably run our pre-built binary distribution instead.

I actually thought of doing something like this in `john.exe`, so it'd work right starting from the second invocation of it - but I'd prefer us finding/implementing a cleaner solution.

**Comment by alainesp** [comment]
> This only works for those building from source on the very machine where they intend to run. Most Windows users will probably run our pre-built binary distribution instead.

I understand. Question. How does JtR manage the distribution of OpenCL-supported JtR on Linux on systems without libOpenCl.so?

The solution that I guess gives the most compatibility is to not link to OpenCL statically, create a *.c file with all the OpenCL functions, and implement them as a hub to the pointer of functions dynamically loaded. If we are lucky the compiler will inline most of them and we probably only need to check if the pointers are good on the couple of OpenCL functions that you can call at the beginning.

**Comment by solardiz** [comment]
> How does JtR manage the distribution of OpenCL-supported JtR on Linux on systems without libOpenCl.so?

It does not, and there's not as much need for that. Packages for Linux typically target a specific distro/version and have dependencies on other packages specified in them. As an alternative for that, @claudioandre-br provides Flatpak and Snap packages, which bundle their dependencies. Anyway, we recommend building for Linux from source, and the releases on JtR homepage are currently only source tarballs except for Windows, where we do provide binary builds.

> The solution that I guess gives the most compatibility is to not link to OpenCL statically, create a `*.c` file with all the OpenCL functions, and implement them as a hub to the pointer of functions dynamically loaded. If we are lucky the compiler will inline most of them and we probably only need to check if the pointers are good on the couple of OpenCL functions that you can call at the beginning.

Right, I think (but am not sure) this is similar to what hashcat does (please take a look). I was thinking we'd use an `.h` file with macros wrapping around `dlsym`, but I don't mind wrapper functions as well (can also be `static inline` ones in an `.h` file). I'm not worried about performance since OpenCL calls shouldn't be frequent (if they are, we have a problem anyway).

So if you're willing to proceed to try and implement this, please go ahead. Thank you very much!

**Comment by solardiz** [comment]
> I think (but am not sure) this is similar to what hashcat does (please take a look).

@alainesp See https://github.com/openwall/john/issues/3191#issuecomment-508119262

**Comment by solardiz** [comment]
> I'm not worried about performance since OpenCL calls shouldn't be frequent (if they are, we have a problem anyway).

That said, we should cache `dlsym` results. The overhead of a function call doesn't worry me much, but the overhead of redoing a symbol lookup by name could be too great.

**Comment by alainesp** [comment]
> > How does JtR manage the distribution of OpenCL-supported JtR on Linux on systems without libOpenCl.so?
> 
> It does not, and there's not as much need for that. Packages for Linux typically target a specific distro/version and have dependencies on other packages specified in them. As an alternative for that, @claudioandre-br provides Flatpak and Snap packages, which bundle their dependencies.

Could you do that? Reference a supposed driver library from a Linux package manager? I was under the impression that the OpenCL support was installed by the GPU driver, at least on Windows.

**Comment by solardiz** [comment]
> Could you do that? Reference a supposed driver library from a Linux package manager?

I think the auto-generated dependency would be on `libOpenCL.so.1` or such, and e.g. on Debian/Ubuntu that can be provided by https://packages.debian.org/sid/ocl-icd-libopencl1 - which is just the loader - or by the actual GPU driver package. You could also add various dependencies manually (such as to require a specific OpenCL backend installed via the package manager), but that's inconvenient and not normally done.

Anyway, I think we should go for the `dlopen` approach for Windows, and when we have it then also use it on Linux (which would also provide more debugging/testing of it).

**Comment by solardiz** [comment]
> I think we should go for the `dlopen` approach for Windows, and when we have it then also use it on Linux

We now have initial implementation of this in #5386 by @alainesp - thank you, Alain!

**Comment by mainmind83** [comment]
> * Locate the vendor's OpenCL dll file. In my case it is [1]:
>   `C:\WINDOWS\System32\DriverStore\FileRepository\u0357645.inf_amd64_3c66802dbd4d45a0\B357435\amdocl64.dll`


If there are several directories with the DLL library, use the one with the most recent date on C:\WINDOWS\System32\DriverStore\FileRepository

It might also be possible to use the [ObjectListView](https://objectlistview.sourceforge.net/cs/index.html) component of
[Driver Store Explore](https://github.com/lostindark/DriverStoreExplorer?tab=readme-ov-file) where it is possible to access the directory with the dll of AMD-OpenCL User Mode Driver:

![image](https://github.com/openwall/john/assets/18678468/95e0a65d-69fb-4cf1-8c4f-a007b25d3d8e)

**Comment by solardiz** [comment]
Thank you @mainmind83, but none of this should be needed now that we've merged @alainesp's #5386 and there's a new Windows build at https://github.com/openwall/john-packages/releases/tag/jumbo-dev also linked from the "Download Windows Build" badge in `README.md` in this repo. So I'll close this issue as completed. We already had some people report that those changes have helped, but it'd be great if you @mainmind83 also test and let us know in here (it's OK to add comments to a closed issue).

**Comment by mainmind83** [comment]
It was still open 🥲 
It's working without edit .icd file but appear warning:

```
john.exe --test --format=nt-opencl
Device 1: gfx1031 [AMD Radeon RX 6700 XT]
Benchmarking: NT-opencl [MD4 OpenCL/mask accel]... Build log: warning: argument unused during compilation: '-I opencl' [-Wunused-command-line-argument]
1 warning generated.

Build log: warning: argument unused during compilation: '-I opencl' [-Wunused-command-line-argument]
1 warning generated.

LWS=128 GWS=327680 (2560 blocks) x9500 DONE
Raw:    36402M c/s real, 6031G c/s virtual
```

**Comment by solardiz** [comment]
Thank you for testing this @mainmind83!

The `Build log: warning: argument unused during compilation: '-I opencl'` is a separate (non-)issue related to our fix of #4335, which was #4373. I'm surprised by this warning, though, since `opencl/nt_kernel.cl` does include a few other files and we're only `chdir`'ing to the `run` directory, not inside `opencl`. Maybe that's an extra reason for us to `chdir` to `run/opencl` instead, like we discussed in https://github.com/openwall/john/pull/4938#issuecomment-992621520

Edit: Upon a second thought, I am not surprised by the warning. Perhaps the newer/better OpenCL runtimes are looking for the `#include "..."` files in the same directory with the file that includes them, like it would be in C. However, we historically had problems with this not happening, which is why we added the `-I` (with a different path back then).

**Comment by sashko-do** [comment]
https://github.com/openwall/john/issues/3191#issuecomment-502644111

> and fortunately, i have resolved the issue by replacing the contents of nvidia.icd under 'etc/OpenCL/vendors'. The default contents of the file is 'c:\Windows\System32\nvopencl.dll'. I have replaced this with the actual path of nvopencl64.dll in another path under system32 folder. Now jon the ripper successfully listed my gpu when using the command 'john --list=opencl-devices'

Worked for me, thanks
